utils/reporter_project_query.py

```diff
@@ -0,0 +1,158 @@
+        help="Space separated list of grant numbers to query NIH RePORTER",
+    )
+    parser.add_argument(
+        "-y",
+        nargs="+",
+        default=[],
+        help="The list of years to use when searching for grant information.",
+    )
+    parser.add_argument("-c", help="Path for CSV where converted JSON will be stored.")
+
+    return parser.parse_args()
+
+
+def build_payload(grant_numbers, years, lim):
+
+    criteria_dict = dict(
+        project_nums=grant_numbers, sub_project_only=True, fiscal_years=years
+    )
+
+    include = [
+        "ProjectTitle",
+        "ProjectNum",
+        "SubprojectId",
+        "FiscalYear",
+        "ProjectEndDate",
+    ]
+
+    payload = dict(
+        criteria=criteria_dict,
+        include_fields=include,
+        offset=0,
+        limit=lim,
+        sort_field="fiscal_year",
+        sort_order="desc",
+    )
+
+    json_payload = json.dumps(payload)
+
+    return json_payload
+
+
+def get_reporter_info(search_payload, header_content):
+
+    req = requests.post(
+        url="https://api.reporter.nih.gov/v2/projects/search",
+        headers=header_content,
+        data=search_payload,
+    )
+
+    return req
+
+
+def build_report(report_list, out_path):
+
+    df_list = []
+    for report in report_list:
+        report_df = pd.read_json(report, orient="records")
+        df_list.append(report_df)
+
+    full_report = pd.concat(df_list).reset_index(drop=True)
+
+    print(f"\n\nYour report is available at {out_path}.\n\n")
+
+    full_report.to_csv(out_path, index=False)
+
+
+def main():
+
+    args = get_args()
+
+    in_grants, in_years, csvPath = args.g, args.y, args.c
+
+    grants = []
+
+    years = []
+
+    for grant in in_grants:
+        grant = "".join([grant, "*"])
+        grants.append(grant)
+
+    for year in in_years:
+        year = int(year)
+        years.append(year)
+
+    headers = {"content-type": "application/json"}
+
+    max_request_length = int(400)
+    grant_count = int(len(grants))
+    req_per_grant = int(20)
+
+    if grant_count > max_request_length / req_per_grant:
+        split_count = math.ceil(grant_count / req_per_grant) + 1
+        print(
+            f"Due to API restrictions, your query will be split into {split_count} parts"
+        )
+        grant_chunks = [g.tolist() for g in np.array_split(grants, split_count)]
+
+    else:
+        grant_chunks = grants
+
+    all_reports = []
+
+    for grant_list in grant_chunks:
+
+        print(f"Building your RePORTER query...")
+        query = build_payload(
+            grant_numbers=grant_list, years=years, lim=max_request_length
+        )
+        valid = json.loads(query)
+
+        if valid:
+            print(f"\n\nSUCCESS! Your query was converted to a valid JSON string")
+
+        else:
+            print(
+                f"\n\nNot quite! Your query: \n {query} \n is an invalid JSON string.\nPlease revise your query and try again."
+            )
+            break
```

Comment:
I wonder if this check is needed?  Because `build_payload()` returns a JSON string, my assumption would be that `json.loads(query)` is always truth-y, therefore, is always valid? Unless there's a use-case I'm not considering.... (it _is_ Friday afternoon, after all 😎 )

Reply:
You're right, this is not needed! I had that in there when I was evaluating ways to build JSON queries, but I moved the final `json_payload = json.dumps(payload)` into `build_payload` and neglected to remove the `json.loads` call in `main`

Reply:
Ah, makes sense!